### PR TITLE
Dynamic Casting: Properly unwrap existential metatype sources

### PR DESCRIPTION
**Pull request by tbkka**
Existential metatypes are really just existentials that hold metatypes.  As
such, they should be handled in the general casting logic in much the same way
as regular existentials: They should generally be ignored by most casting logic,
and unwrapped as necessary at the top level.

In particular, the previous code would fail to correctly handle the following
cast from an existential metatype (`AnyObject.Type`) to an existential
(`AnyObject`):
```
  class C {}
  let a = C.self as AnyObject.Type
  let b = a as! AnyObject
```
With the old code, `b` above would hold a reference to a `__SwiftValue` box
containing the type reference.  The correct result would simply store the type
reference directly in `b`.  These two are only really distinguishable in that
the correct form permits `a === b` to return `true`.

Fixes rdar://70582753

### CODE REVIEW

**Comment by tbkka** [comment]
@swift-ci Please test

**Review by mikeash — APPROVED** [review]

File: test/Casting/Casts.swift [hunk 1]
```diff
@@ -761,4 +761,27 @@ CastsTests.test("Async function types") {
...
+  let e = c as? C.Type
+  expectNotNil(e)
+  expectTrue(e! === a)
+  let f = runtimeCast(d, to: C.Type.self)
+  expectNotNil(f)
+  expectTrue(f! === a)
+  blackhole(c)
+  blackhole(d)
+  blackhole(e)
+  blackhole(f)
```
[the remark this answers is not in the document]
Comment (tbkka) [reply test/Casting/Casts.swift]: In fact they do not.  Removed.

**Comment by tbkka** [comment]
@swift-ci Please test

**Comment by tbkka** [comment]
@swift-ci Please test

**Pull request merged**